src/main/java/com/sequenceiq/cloudbreak/service/stack/StackFailureHandlerService.java

```diff
@@ -0,0 +1,143 @@
+                LOGGER.info("Failure policy is null so error will throw");
+                throwError(stack, resourceRequestResults);
+            }
+        } else {
+            switch (stack.getFailurePolicy().getAdjustmentType()) {
+                case EXACT:
+                    if (stack.getFailurePolicy().getThreshold() < resourceRequestResults.size()) {
+                        LOGGER.info("Number of failures is more than the threshold so error will throw");
+                        throwError(stack, resourceRequestResults);
+                    } else if (resourceRequestResults.size() != 0) {
+                        LOGGER.info("Decrease node counts because threshold was higher");
+                        handleExceptions(stack, resourceRequestResults);
+                    }
+                    break;
+                case PERCENTAGE:
+                    if (Double.valueOf(stack.getFailurePolicy().getThreshold())
+                            < Double.valueOf(resourceRequestResults.size()) / Double.valueOf(stack.getFullNodeCount()) * ONE_HUNDRED) {
+                        LOGGER.info("Number of failures is more than the threshold so error will throw");
+                        throwError(stack, resourceRequestResults);
+                    } else if (resourceRequestResults.size() != 0) {
+                        LOGGER.info("Decrease node counts because threshold was higher");
+                        handleExceptions(stack, resourceRequestResults);
+                    }
+                    break;
+                default:
+                    throwError(stack, resourceRequestResults);
+                    break;
+            }
+        }
+    }
+
+    private void handleExceptions(Stack stack, List<ResourceRequestResult> resourceRequestResults) {
+        for (ResourceRequestResult exception : resourceRequestResults) {
+            List<Resource> resourceList = new ArrayList<>();
+            for (Resource resource : exception.getResources()) {
+                Resource newResource = resourceRepository.findByStackIdAndName(stack.getId(), resource.getResourceName(), resource.getResourceType());
+                if (newResource != null) {
+                    resourceList.add(newResource);
+                }
+            }
+            if (!resourceList.isEmpty()) {
+                doRollbackAndDecreaseNodeCount(exception, stack, resourceList, resourceRequestResults);
+            }
+        }
+    }
+
+    private void doRollbackAndDecreaseNodeCount(ResourceRequestResult exception, Stack stack, List<Resource> resourceList,
+            List<ResourceRequestResult> resourceRequestResults) {
+        InstanceGroup instanceGroup = instanceGroupRepository.findOne(exception.getInstanceGroup().getId());
+        instanceGroup.setNodeCount(instanceGroup.getNodeCount() - 1);
+        if (instanceGroup.getNodeCount() <= 0) {
+            LOGGER.info("InstanceGroup node count lower than 1 which is incorrect so error will throw");
```

Comment:
Just ran a test on Azure, worked well. 1 more test tomorrow and I'm fine with it, but in the meanwhile could you please extend the logging with more information please? Which hosts group's node failed, why, from what number are we decreasing to what with mdc context fixed. Also log which "rule" applied exact or percent what is the threshold. It will help a lot later.